vm/systemSmartContracts/governance.go

```diff
@@ -147,73 +165,268 @@ func (g *governanceContract) init(args *vmcommon.ContractCallInput) vmcommon.Ret
 	}
 
-	scConfig.NumNodes = numNodes.Int64()
-	scConfig.MinQuorum = int32(minQuorum.Int64())
-	scConfig.MinVetoThreshold = int32(minVeto.Int64())
-	scConfig.MinPassThreshold = int32(minPass.Int64())
+	currentVoteData, err := g.getOrCreateVoteData(proposalToVote, voterAddress)
+	if err != nil {
+		g.eei.AddReturnMessage(err.Error())
+		return vmcommon.ExecutionFailed
+	}
+	if totalVotingPower.Cmp(big.NewInt(0).Add(votePower, currentVoteData.UsedPower)) == -1 {
+		g.eei.AddReturnMessage("not enough voting power to cast this vote")
+		return vmcommon.UserError
+	}
 
-	marshaledData, err := g.marshalizer.Marshal(scConfig)
+	currentVote := &VoteDetails{
+		Value:       voteOption,
+		Power:       votePower,
+		DelegatedTo: delegatedTo,
+		Balance:     big.NewInt(0),
+		Type:        Validator,
+	}
+	newVoteData, updatedProposal, err := g.applyVote(currentVote, currentVoteData, proposal)
 	if err != nil {
-		g.eei.AddReturnMessage("changeConfig error " + err.Error())
+		g.eei.AddReturnMessage(err.Error())
 		return vmcommon.UserError
 	}
-	g.eei.SetStorage([]byte(governanceConfigKey), marshaledData)
+
+	err = g.saveNewVoteData(voterAddress, newVoteData, updatedProposal, Validator)
+	if err != nil {
+		g.eei.AddReturnMessage(err.Error())
+		return vmcommon.ExecutionFailed
+	}
 
 	return vmcommon.Ok
 }
 
-func (g *governanceContract) getConfig() (*GovernanceConfig, error) {
-	marshaledData := g.eei.GetStorage([]byte(governanceConfigKey))
-	scConfig := &GovernanceConfig{}
-	err := g.marshalizer.Unmarshal(scConfig, marshaledData)
+// claimFunds returns back the used funds for a particular proposal if they are unlocked. Accepts a single parameter:
+//  args.Arguments[0] - proposal reference
+func (g *governanceContract) claimFunds(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	if args.CallValue.Cmp(big.NewInt(0)) != 0 {
+		g.eei.AddReturnMessage("invalid callValue, should be 0")
+		return vmcommon.UserError
+	}
+
+	if len(args.Arguments) != 1 {
+		g.eei.AddReturnMessage("invalid number of arguments, expected 1")
+		return vmcommon.FunctionWrongSignature
+	}
+
+	lock := g.getLock(args.CallerAddr, Account, args.Arguments[0])
+	currentNonce := g.eei.BlockChainHook().CurrentNonce()
+
+	if lock.Cmp(big.NewInt(0).SetUint64(currentNonce)) == -1 {
+		g.eei.AddReturnMessage("your funds are still locked")
+		return vmcommon.UserError
+	}
+
+	currentVoteData, err := g.getOrCreateVoteData(args.Arguments[0], args.CallerAddr)
 	if err != nil {
-		return nil, err
+		g.eei.AddReturnMessage(err.Error())
+		return vmcommon.ExecutionFailed
+	}
+	if currentVoteData.Claimed == true {
+		g.eei.AddReturnMessage("you already claimed back your funds")
+		return vmcommon.UserError
 	}
 
-	return scConfig, nil
+	fundsToTransfer := big.NewInt(0)
+	for _, vote := range currentVoteData.VoteItems {
+		if vote.Type != Account {
+			continue
+		}
+
+		fundsToTransfer.Add(fundsToTransfer, vote.Balance)
+	}
+
+	if fundsToTransfer.Cmp(big.NewInt(0)) != 1 {
```

Comment:
Done, changed everywhere. I just got a chill, thinking somebody in the future will mistakingly/maliciously  do `zero.Add(x, y)`. Think of it, in a SC public function, not really used by anybody. And it can sit there, until somebody decides to call it :)

Reply:
that is true. Because the big int can not be defined as a const...